// shaders/dist/core/AngularBlur-CuwAjUOA.js
import { s as transformPosition } from "./transformations-YbhRK-rd.js";
import { Fn, convertToTexture, cos, float, screenUV, sin, vec2, vec4 } from "three/tsl";
const componentDefinition = {
	name: "AngularBlur",
	category: "Blurs",
	description: "Radial motion blur rotating around a center point",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 20,
			description: "Intensity of the angular blur effect",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Blur Intensity"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the rotation",
			ui: {
				type: "position",
				label: "Center Position"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const center = vec2(uniforms.center.uniform.x, uniforms.center.uniform.y.oneMinus());
		const angle = uniforms.intensity.uniform.mul(.005);
		return Fn(() => {
			const initialCoord = screenUV.sub(center);
			const total = vec4(0).toVar();
			const totalWeight = float(0).toVar();
			const angleStep = angle.div(float(31));
			for (let i = 0; i < 32; i++) {
				const t = float(i).div(float(31)).sub(.5).mul(2);
				const weight = float(1).div(float(2.506628)).mul(float(-.5).mul(t).mul(t).div(float(.64)).exp());
				const currentAngle = angleStep.mul(float(i));
				const cosAngle = cos(currentAngle);
				const sinAngle = sin(currentAngle);
				const sampleCoord = vec2(initialCoord.x.mul(cosAngle).sub(initialCoord.y.mul(sinAngle)), initialCoord.x.mul(sinAngle).add(initialCoord.y.mul(cosAngle))).add(center);
				const sample = childTexture.sample(sampleCoord).mul(weight);
				total.assign(total.add(sample));
				totalWeight.assign(totalWeight.add(weight));
			}
			return total.div(totalWeight);
		})();
	}
};
var AngularBlur_default = componentDefinition;
export { componentDefinition as n, AngularBlur_default as t };
