// shaders/dist/core/Blur-k37gOaRA.js
import { Fn, convertToTexture, float, screenUV, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Blur",
	category: "Blurs",
	description: "A simple Gaussian blur effect",
	requiresRTT: true,
	requiresChild: true,
	props: { intensity: {
		default: 50,
		description: "Intensity of the blur effect",
		ui: {
			type: "range",
			min: 0,
			max: 100,
			step: 1,
			label: "Intensity"
		}
	} },
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const weights = [
			.056,
			.135,
			.278,
			.487,
			.726,
			.923,
			1,
			.923,
			.726,
			.487,
			.278,
			.135,
			.056
		];
		const weightSum = 6.214;
		const scaledIntensity = uniforms.intensity.uniform.mul(.06);
		const horizontalTexture = convertToTexture(Fn(() => {
			const uv$1 = screenUV;
			const blurRadius = scaledIntensity;
			const pixelSize = vec2(1).div(viewportSize);
			const total = vec4(0).toVar();
			for (let x = -6; x <= 6; x++) {
				const weight = float(weights[x + 6] / weightSum);
				const offset = vec2(float(x), float(0)).mul(blurRadius).mul(pixelSize);
				const sampleCoord = uv$1.add(offset);
				const sample = childTexture.sample(sampleCoord).mul(weight);
				total.assign(total.add(sample));
			}
			return total;
		})());
		onCleanup(() => {
			if (horizontalTexture?.renderTarget?.dispose) horizontalTexture.renderTarget.dispose();
		});
		return Fn(() => {
			const uv$1 = screenUV;
			const blurRadius = scaledIntensity;
			const pixelSize = vec2(1).div(viewportSize);
			const total = vec4(0).toVar();
			for (let y = -6; y <= 6; y++) {
				const weight = float(weights[y + 6] / weightSum);
				const offset = vec2(float(0), float(y)).mul(blurRadius).mul(pixelSize);
				const sampleCoord = uv$1.add(offset);
				const sample = horizontalTexture.sample(sampleCoord).mul(weight);
				total.assign(total.add(sample));
			}
			return total;
		})();
	}
};
var Blur_default = componentDefinition;
export { componentDefinition as n, Blur_default as t };
