// shaders/dist/core/Circle-BlVzkeyl.js
import { r as transformColor, s as transformPosition } from "./transformations-YbhRK-rd.js";
import { length, screenUV, smoothstep, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Circle",
	category: "Base Layers",
	description: "Generate a circle with adjustable size and softness",
	props: {
		color: {
			default: "#ffffff",
			transform: transformColor,
			description: "The color of the circle",
			ui: {
				type: "color",
				label: "Color"
			}
		},
		radius: {
			default: 1,
			description: "The radius of the circle. A value of one (1) is sets the circle to fit the canvas.",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .01,
				label: "Radius"
			}
		},
		softness: {
			default: 0,
			description: "Edge softness. Lower values like zero (0) are sharp, higher values like one (1) are softer.",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Softness"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the circle",
			ui: {
				type: "position",
				label: "Center Position"
			}
		}
	},
	fragmentNode: ({ uniforms }) => {
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectCorrectedUV = vec2(uv$1.x.mul(aspect), uv$1.y);
		const centerPos = vec2(uniforms.center.uniform.x.mul(aspect), uniforms.center.uniform.y.oneMinus());
		const distanceFromCenter = length(aspectCorrectedUV.sub(centerPos));
		const edgeSoftness = uniforms.softness.uniform;
		const circleEdge = uniforms.radius.uniform.mul(.5);
		const circleMask = smoothstep(circleEdge, circleEdge.sub(edgeSoftness), distanceFromCenter);
		return vec4(uniforms.color.uniform.rgb, uniforms.color.uniform.a.mul(circleMask));
	}
};
var Circle_default = componentDefinition;
export { componentDefinition as n, Circle_default as t };
