// shaders/dist/core/Dither-C1KyBYuE.js
import { r as transformColor } from "./transformations-YbhRK-rd.js";
import { dot, floor, mix, screenUV, step, vec2, vec3, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Dither",
	category: "Stylize",
	description: "Bayer matrix dithering to reduce colors",
	requiresChild: true,
	props: {
		pixelSize: {
			default: 4,
			description: "Size of dithering pixels",
			ui: {
				type: "range",
				min: 1,
				max: 20,
				step: 1,
				label: "Pixel Size"
			}
		},
		threshold: {
			default: .5,
			description: "Luminance threshold for dithering",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Threshold"
			}
		},
		colorA: {
			default: "transparent",
			transform: transformColor,
			description: "Dark color for dithering",
			ui: {
				type: "color",
				label: "Dark Color"
			}
		},
		colorB: {
			default: "#ffffff",
			transform: transformColor,
			description: "Light color for dithering",
			ui: {
				type: "color",
				label: "Light Color"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const sourceColor = childNode;
		const luminance = dot(sourceColor.rgb, vec3(.299, .587, .114)).mul(sourceColor.a);
		const pixelSize = uniforms.pixelSize.uniform;
		const bayerPos = floor(screenUV.mul(viewportSize).div(pixelSize)).mod(vec2(4));
		const x = bayerPos.x;
		const y = bayerPos.y;
		const bayerValue = y.equal(0).select(x.equal(0).select(0 / 16, x.equal(1).select(8 / 16, x.equal(2).select(2 / 16, 10 / 16))), y.equal(1).select(x.equal(0).select(12 / 16, x.equal(1).select(4 / 16, x.equal(2).select(14 / 16, 6 / 16))), y.equal(2).select(x.equal(0).select(3 / 16, x.equal(1).select(11 / 16, x.equal(2).select(1 / 16, 9 / 16))), x.equal(0).select(15 / 16, x.equal(1).select(7 / 16, x.equal(2).select(13 / 16, 5 / 16))))));
		const threshold = uniforms.threshold.uniform;
		const ditherResult = step(bayerValue, luminance.add(threshold.sub(.5)));
		return vec4(mix(uniforms.colorA.uniform.rgb, uniforms.colorB.uniform.rgb, ditherResult), mix(uniforms.colorA.uniform.a, uniforms.colorB.uniform.a, ditherResult));
	}
};
var Dither_default = componentDefinition;
export { componentDefinition as n, Dither_default as t };
