// shaders/dist/core/DotGrid-ll8XCA0p.js
import { r as transformColor } from "./transformations-YbhRK-rd.js";
import { float, floor, fwidth, length, mix, screenUV, sin, smoothstep, time, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "DotGrid",
	category: "Base Layers",
	description: "Grid of dots with optional twinkling animation",
	props: {
		color: {
			default: "#ffffff",
			description: "The color of the dot",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color"
			}
		},
		density: {
			default: 100,
			description: "The number of dots on the longest canvas edge",
			ui: {
				type: "range",
				min: 10,
				max: 500,
				step: 10,
				label: "Density"
			}
		},
		dotSize: {
			default: .3,
			description: "The size of each dot, zero (0) being invisible, one (1) filled the grid with no gaps",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Dot Size"
			}
		},
		twinkle: {
			default: 0,
			description: "Intensity of the twinkle effect (0 = off, 1 = full twinkle)",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Twinkle"
			}
		}
	},
	fragmentNode: ({ uniforms }) => {
		const uvCoords = screenUV;
		const density = uniforms.density.uniform;
		const dotSize = uniforms.dotSize.uniform;
		const dotColor = uniforms.color.uniform;
		const twinkleIntensity = uniforms.twinkle.uniform;
		const resolution = viewportSize;
		const aspect = resolution.x.div(resolution.y);
		const correctedUV = vec2(uvCoords.x.mul(aspect), uvCoords.y.oneMinus());
		const gridUV = correctedUV.mul(density);
		const centerDistance = length(gridUV.sub(floor(gridUV)).sub(.5));
		const pixelSize = length(fwidth(correctedUV.mul(density)));
		const dot$1 = float(1).sub(smoothstep(dotSize.mul(.5), dotSize.mul(.5).add(pixelSize.mul(.5)), centerDistance));
		const gridPosition = floor(gridUV);
		const dotPhase = gridPosition.x.mul(12.9898).add(gridPosition.y.mul(78.233)).sin().mul(43758.5453).fract();
		const twinkleValue = sin(time.mul(2).add(dotPhase.mul(6.28318))).mul(.5).add(.5);
		const twinkleModifier = mix(float(1), twinkleValue, twinkleIntensity);
		const twinkledAlpha = dot$1.mul(twinkleModifier);
		return vec4(dotColor.rgb, twinkledAlpha);
	}
};
var DotGrid_default = componentDefinition;
export { componentDefinition as n, DotGrid_default as t };
