// shaders/dist/core/Glow-BYNKuw0E.js
import { Fn, convertToTexture, dot, float, max, screenUV, vec2, vec3, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Glow",
	category: "Stylize",
	description: "Soft glow effect with adjustable intensity",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 1,
			description: "Glow intensity (brightness of the glow effect)",
			ui: {
				type: "range",
				min: 0,
				max: 10,
				step: .01,
				label: "Intensity"
			}
		},
		threshold: {
			default: .5,
			description: "Brightness threshold for glow extraction (lower = more glow)",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Threshold"
			}
		},
		size: {
			default: 10,
			description: "Size/spread of the glow effect",
			ui: {
				type: "range",
				min: 1,
				max: 20,
				step: .01,
				label: "Glow Size"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const originalColor = childTexture.sample(screenUV);
		const glowEffect = Fn(() => {
			const blurRadius = uniforms.size.uniform;
			const threshold = uniforms.threshold.uniform;
			const pixelSize = vec2(1).div(viewportSize);
			const blurredGlow = vec4(0, 0, 0, 0).toVar();
			const radius = 5;
			for (let dx = -radius; dx <= radius; dx++) for (let dy = -radius; dy <= radius; dy++) {
				const offset = vec2(float(dx), float(dy)).mul(pixelSize).mul(blurRadius);
				const sampleUV = screenUV.add(offset);
				const sampleColor = childTexture.sample(sampleUV);
				const luminance = dot(sampleColor.rgb, vec3(.299, .587, .114));
				const brightnessMask = max(float(0), luminance.sub(threshold));
				const brightColor = sampleColor.mul(brightnessMask);
				blurredGlow.assign(blurredGlow.add(brightColor));
			}
			const totalSamples = float((radius * 2 + 1) * (radius * 2 + 1));
			return blurredGlow.div(totalSamples);
		})();
		const intensity = uniforms.intensity.uniform;
		const intensifiedGlow = glowEffect.mul(intensity);
		return vec4(originalColor.rgb.add(intensifiedGlow.rgb), max(originalColor.a, intensifiedGlow.a));
	}
};
var Glow_default = componentDefinition;
export { componentDefinition as n, Glow_default as t };
