// shaders/dist/core/Grid-CdslWvoH.js
import { r as transformColor } from "./transformations-YbhRK-rd.js";
import { fract, fwidth, max, screenUV, smoothstep, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Grid",
	category: "Base Layers",
	description: "Simple grid lines pattern with adjustable thickness",
	props: {
		color: {
			default: "#ffffff",
			transform: transformColor,
			description: "The color of the grid lines",
			ui: {
				type: "color",
				label: "Color"
			}
		},
		cells: {
			default: 10,
			description: "Number of cells along the shortest canvas edge (creates square cells)",
			ui: {
				type: "range",
				min: 1,
				max: 50,
				step: 1,
				label: "Cells"
			}
		},
		thickness: {
			default: 1,
			description: "Thickness of grid lines (normalized, 0.0-1.0)",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Thickness"
			}
		}
	},
	fragmentNode: ({ uniforms }) => {
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const correctedUV = vec2(uv$1.x.mul(aspect), uv$1.y.oneMinus());
		const cells = uniforms.cells.uniform;
		const gridUV = correctedUV.mul(cells);
		const gridFract = fract(gridUV);
		const lineWidth = uniforms.thickness.uniform.mul(.01);
		const distToLineX = gridFract.x.min(gridFract.x.oneMinus());
		const distToLineY = gridFract.y.min(gridFract.y.oneMinus());
		const pixelSize = fwidth(gridUV);
		const gridMask = max(smoothstep(lineWidth.add(pixelSize.x), lineWidth, distToLineX), smoothstep(lineWidth.add(pixelSize.y), lineWidth, distToLineY));
		return vec4(uniforms.color.uniform.rgb, uniforms.color.uniform.a.mul(gridMask));
	}
};
var Grid_default = componentDefinition;
export { componentDefinition as n, Grid_default as t };
