// shaders/dist/core/Halftone-CV0aFuRG.js
import { n as transformAngle } from "./transformations-YbhRK-rd.js";
import { cos, dot, fract, length, radians, screenUV, sin, smoothstep, vec2, vec3, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Halftone",
	category: "Stylize",
	description: "Halftone dot pattern effect for printing aesthetics",
	requiresChild: true,
	props: {
		frequency: {
			default: 100,
			description: "Frequency of the halftone dots",
			ui: {
				type: "range",
				min: 10,
				max: 300,
				step: 1,
				label: "Frequency"
			}
		},
		angle: {
			default: 45,
			transform: transformAngle,
			description: "Rotation angle of the pattern (in degrees)",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Angle"
			}
		},
		smoothness: {
			default: .1,
			description: "Edge smoothness of the dots",
			ui: {
				type: "range",
				min: 0,
				max: .3,
				step: .01,
				label: "Smoothness"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode }) => {
		if (!childNode) return vec4(0);
		const _uv = screenUV;
		const frequency = uniforms.frequency.uniform;
		const angleDegrees = uniforms.angle.uniform;
		const smoothness = uniforms.smoothness.uniform;
		const angle = radians(angleDegrees);
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectCorrectedUV = vec2(_uv.x.mul(aspect), _uv.y);
		const c = cos(angle);
		const s = sin(angle);
		const gridUV = fract(vec2(dot(aspectCorrectedUV, vec2(c, s.negate())), dot(aspectCorrectedUV, vec2(s, c))).mul(frequency)).sub(.5);
		const originalColor = childNode;
		const dotSize = dot(originalColor.rgb, vec3(.299, .587, .114)).mul(.7).add(.15);
		const dist = length(gridUV);
		const dotPattern = smoothstep(dotSize.add(smoothness), dotSize.sub(smoothness), dist);
		return vec4(originalColor.rgb.mul(dotPattern), originalColor.a.mul(dotPattern));
	}
};
var Halftone_default = componentDefinition;
export { componentDefinition as n, Halftone_default as t };
