// shaders/dist/core/LinearBlur-CTXjolzL.js
import { n as transformAngle } from "./transformations-YbhRK-rd.js";
import { Fn, convertToTexture, cos, float, radians, screenUV, sin, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "LinearBlur",
	category: "Blurs",
	description: "Directional motion blur in a specific angle",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 30,
			description: "Intensity of the linear blur effect",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Direction of the linear blur (in degrees)",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Angle"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const angleRad = radians(uniforms.angle.uniform);
		const blurDirection = vec2(cos(angleRad), sin(angleRad));
		return Fn(() => {
			const uv$1 = screenUV;
			const texel = vec2(1).div(viewportSize);
			const blurVector = blurDirection.mul(uniforms.intensity.uniform).mul(texel).mul(2);
			const total = vec4(0).toVar();
			const totalWeight = float(0).toVar();
			for (let i = 0; i < 32; i++) {
				const t = float(i).div(float(31)).sub(.5).mul(2);
				const weight = float(1).div(float(2.506628)).mul(float(-.5).mul(t).mul(t).div(float(.64)).exp());
				const coord = uv$1.add(blurVector.mul(t.mul(.5)));
				const sample = childTexture.sample(coord).mul(weight);
				total.assign(total.add(sample));
				totalWeight.assign(totalWeight.add(weight));
			}
			return total.div(totalWeight);
		})();
	}
};
var LinearBlur_default = componentDefinition;
export { componentDefinition as n, LinearBlur_default as t };
