// shaders/dist/core/ProgressiveBlur-RuAm_BKx.js
import { n as transformAngle, s as transformPosition } from "./transformations-YbhRK-rd.js";
import { Fn, convertToTexture, cos, float, max, radians, screenUV, sin, smoothstep, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "ProgressiveBlur",
	category: "Blurs",
	description: "Blur that increases progressively in one direction",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 50,
			description: "Maximum intensity of the blur effect",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity"
			}
		},
		angle: {
			default: 0,
			transform: transformAngle,
			description: "Direction of the blur gradient (in degrees)",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Angle"
			}
		},
		center: {
			default: {
				x: 0,
				y: .5
			},
			transform: transformPosition,
			description: "Center point where blur begins",
			ui: {
				type: "position",
				label: "Center"
			}
		},
		falloff: {
			default: 1,
			description: "Distance over which blur transitions to full strength",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Falloff"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const angleRad = radians(uniforms.angle.uniform);
		const direction = vec2(cos(angleRad), sin(angleRad));
		const centerPosition = vec2(uniforms.center.uniform[0], uniforms.center.uniform[1].oneMinus());
		const centeredUV = screenUV.sub(centerPosition);
		const directionalDist = max(float(0), centeredUV.dot(direction));
		const falloff = max(float(.001), uniforms.falloff.uniform);
		const blurAmount = smoothstep(float(0), falloff, directionalDist);
		const scaledIntensity = uniforms.intensity.uniform.mul(.06);
		const blurRadius = blurAmount.mul(scaledIntensity);
		const weights = [
			.056,
			.135,
			.278,
			.487,
			.726,
			.923,
			1,
			.923,
			.726,
			.487,
			.278,
			.135,
			.056
		];
		const weightSum = 6.214;
		const horizontalTexture = convertToTexture(Fn(() => {
			const uv$1 = screenUV;
			const pixelSize = vec2(1).div(viewportSize);
			const total = vec4(0).toVar();
			for (let x = -6; x <= 6; x++) {
				const weight = float(weights[x + 6] / weightSum);
				const offset = vec2(float(x), float(0)).mul(blurRadius).mul(pixelSize);
				const sampleCoord = uv$1.add(offset);
				const sample = childTexture.sample(sampleCoord).mul(weight);
				total.assign(total.add(sample));
			}
			return total;
		})());
		onCleanup(() => {
			if (horizontalTexture?.renderTarget?.dispose) horizontalTexture.renderTarget.dispose();
		});
		return Fn(() => {
			const uv$1 = screenUV;
			const pixelSize = vec2(1).div(viewportSize);
			const total = vec4(0).toVar();
			for (let y = -6; y <= 6; y++) {
				const weight = float(weights[y + 6] / weightSum);
				const offset = vec2(float(0), float(y)).mul(blurRadius).mul(pixelSize);
				const sampleCoord = uv$1.add(offset);
				const sample = horizontalTexture.sample(sampleCoord).mul(weight);
				total.assign(total.add(sample));
			}
			return total;
		})();
	}
};
var ProgressiveBlur_default = componentDefinition;
export { componentDefinition as n, ProgressiveBlur_default as t };
