// shaders/dist/core/RadialGradient-BfFuRbIl.js
import { i as transformColorSpace, r as transformColor, s as transformPosition, t as colorSpaceOptions } from "./transformations-YbhRK-rd.js";
import { t as mixColors } from "./colorMixing-CZPFmiT4.js";
import { distance, float, screenUV, vec2, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "RadialGradient",
	category: "Base Layers",
	description: "Radial gradient radiating from a center point",
	props: {
		colorA: {
			default: "#ff0000",
			transform: transformColor,
			description: "The starting color at the center of the gradient",
			ui: {
				type: "color",
				label: "Color A"
			}
		},
		colorB: {
			default: "#0000ff",
			transform: transformColor,
			description: "The ending color at the edge of the gradient",
			ui: {
				type: "color",
				label: "Color B"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the radial gradient",
			ui: {
				type: "position",
				label: "Center"
			}
		},
		radius: {
			default: 1,
			description: "The radius of the gradient (normalized, 0.0-1.0)",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .01,
				label: "Radius"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space"
			}
		}
	},
	fragmentNode: ({ uniforms }) => {
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const t = distance(vec2(uv$1.x.mul(aspect), uv$1.y), vec2(uniforms.center.uniform.x.mul(aspect), uniforms.center.uniform.y.oneMinus())).div(uniforms.radius.uniform).clamp(float(0), float(1));
		return mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, t, uniforms.colorSpace.uniform);
	}
};
var RadialGradient_default = componentDefinition;
export { componentDefinition as n, RadialGradient_default as t };
