// shaders/dist/core/RectangularCoordinates-CiW3J4UT.js
import { t as applyEdgeHandling } from "./edges-Bd7GP4s2.js";
import { o as transformEdges, s as transformPosition } from "./transformations-YbhRK-rd.js";
import { PI, convertToTexture, cos, mix, screenUV, sin, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "RectangularCoordinates",
	category: "Distortions",
	description: "Convert polar coordinates back to rectangular space",
	requiresRTT: true,
	requiresChild: true,
	props: {
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point for rectangular coordinate conversion",
			ui: {
				type: "position",
				label: "Center"
			}
		},
		scale: {
			default: 1,
			description: "Scale factor for the rectangular output",
			ui: {
				type: "range",
				min: .1,
				max: 3,
				step: .01,
				label: "Scale"
			}
		},
		intensity: {
			default: 1,
			description: "Blends between original UVs (0) and rectangular coordinates (1)",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Intensity"
			}
		},
		edges: {
			default: "transparent",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const theta = screenUV.x.mul(PI.mul(2)).sub(PI);
		const r = screenUV.y.mul(uniforms.scale.uniform);
		const rectX = r.mul(cos(theta));
		const rectY = r.mul(sin(theta));
		const aspect = viewportSize.x.div(viewportSize.y);
		const centerPos = vec2(uniforms.center.uniform.x, uniforms.center.uniform.y.oneMinus());
		const blendedUV = mix(screenUV, vec2(rectX.div(aspect).add(centerPos.x), rectY.add(centerPos.y)), uniforms.intensity.uniform);
		return applyEdgeHandling(blendedUV, childTexture.sample(blendedUV), childTexture, uniforms.edges.uniform);
	}
};
var RectangularCoordinates_default = componentDefinition;
export { componentDefinition as n, RectangularCoordinates_default as t };
