// shaders/dist/core/registry.js
import "./edges-Bd7GP4s2.js";
import "./transformations-YbhRK-rd.js";
import "./time-BZqyVJXt.js";
import { n as componentDefinition } from "./AngularBlur-CuwAjUOA.js";
import { n as componentDefinition$1 } from "./Ascii-Dxk6VaBp.js";
import "./colorMixing-CZPFmiT4.js";
import { n as componentDefinition$2 } from "./Beam-CbG4MkLo.js";
import { n as componentDefinition$3 } from "./Blob-Ck1tVOL_.js";
import { n as componentDefinition$4 } from "./Blur-k37gOaRA.js";
import { n as componentDefinition$5 } from "./BrightnessContrast-CaErceqR.js";
import { n as componentDefinition$6 } from "./Bulge-BVcWo6On.js";
import { n as componentDefinition$7 } from "./ChannelBlur-CQoP7Igg.js";
import { n as componentDefinition$8 } from "./Checkerboard-D_hGvoO7.js";
import { n as componentDefinition$9 } from "./ChromaFlow-DizYQSgf.js";
import { n as componentDefinition$10 } from "./ChromaticAberration-Bk_DMYJm.js";
import { n as componentDefinition$11 } from "./Circle-BlVzkeyl.js";
import { n as componentDefinition$12 } from "./CRTScreen-CcAYeqoI.js";
import { n as componentDefinition$13 } from "./CursorTrail-CxXWV7KH.js";
import { n as componentDefinition$14 } from "./DiffuseBlur-BTSRl65K.js";
import { n as componentDefinition$15 } from "./Dither-C1KyBYuE.js";
import { n as componentDefinition$16 } from "./DotGrid-ll8XCA0p.js";
import { n as componentDefinition$17 } from "./Duotone-d2gQJr4M.js";
import { n as componentDefinition$18 } from "./FilmGrain-3h6TIe1T.js";
import { n as componentDefinition$19 } from "./FloatingParticles-D95-rj4h.js";
import { n as componentDefinition$20 } from "./GlassTiles-DqT67iqr.js";
import { n as componentDefinition$21 } from "./Glow-BYNKuw0E.js";
import { n as componentDefinition$22 } from "./Godrays-DCk1c8sQ.js";
import { n as componentDefinition$23 } from "./Grayscale-BH3tLDGN.js";
import { n as componentDefinition$24 } from "./Grid-CdslWvoH.js";
import { n as componentDefinition$25 } from "./GridDistortion-D-iiGELF.js";
import { n as componentDefinition$26 } from "./Group-ByGv48Lm.js";
import { n as componentDefinition$27 } from "./Halftone-CV0aFuRG.js";
import { n as componentDefinition$28 } from "./HueShift-DQ2WgImx.js";
import { n as componentDefinition$29 } from "./ImageTexture-DKFb8ze3.js";
import { n as componentDefinition$30 } from "./Invert-w9duGaUu.js";
import { n as componentDefinition$31 } from "./LinearBlur-CTXjolzL.js";
import { n as componentDefinition$32 } from "./LinearGradient-CZo6e7r8.js";
import { n as componentDefinition$33 } from "./Liquify-BbfDBMNZ.js";
import { n as componentDefinition$34 } from "./Perspective-CSP2Hb16.js";
import { n as componentDefinition$35 } from "./Pixelate-BAoi-CSx.js";
import { n as componentDefinition$36 } from "./PolarCoordinates-BwN9HhFK.js";
import { n as componentDefinition$37 } from "./Posterize-XoAqPWv2.js";
import { n as componentDefinition$38 } from "./ProgressiveBlur-RuAm_BKx.js";
import { n as componentDefinition$39 } from "./RadialGradient-BfFuRbIl.js";
import { n as componentDefinition$40 } from "./RectangularCoordinates-CiW3J4UT.js";
import { n as componentDefinition$41 } from "./Ripples-DKBIp3Qq.js";
import { n as componentDefinition$42 } from "./Saturation-DVVlpikK.js";
import { n as componentDefinition$43 } from "./Sharpness-cw57bW0s.js";
import { n as componentDefinition$44 } from "./Shatter-COjV2jLy.js";
import { n as componentDefinition$45 } from "./SimplexNoise-Ce79jgYM.js";
import { n as componentDefinition$46 } from "./SineWave-Cv2gOU8a.js";
import { n as componentDefinition$47 } from "./SolidColor-DFkDBFq3.js";
import { n as componentDefinition$48 } from "./Spherize-R-s4pX9t.js";
import { n as componentDefinition$49 } from "./Spiral-W4dALX1z.js";
import { n as componentDefinition$50 } from "./Strands-DtN5POfO.js";
import { n as componentDefinition$51 } from "./Stretch-BzvP_hdu.js";
import { n as componentDefinition$52 } from "./Stripes-BbNFX05u.js";
import { n as componentDefinition$53 } from "./Swirl-CfsYsr4X.js";
import { n as componentDefinition$54 } from "./TiltShift-B6gbm7Z8.js";
import { n as componentDefinition$55 } from "./Tint-CpnitsHw.js";
import { n as componentDefinition$56 } from "./Tritone-DzvSKVfR.js";
import { n as componentDefinition$57 } from "./Twirl-BXH4uAd6.js";
import { n as componentDefinition$58 } from "./Vibrance-ByN9bRgU.js";
import { n as componentDefinition$59 } from "./WaveDistortion-BA1PyqYH.js";
import { n as componentDefinition$60 } from "./WebcamTexture-D4Uv7Plz.js";
import { n as componentDefinition$61 } from "./ZoomBlur-CIkkDBIo.js";
var createShaderRegistry = () => {
	const registry = {};
	const shaderDefinitions = {
		AngularBlur: componentDefinition,
		Ascii: componentDefinition$1,
		Beam: componentDefinition$2,
		Blob: componentDefinition$3,
		Blur: componentDefinition$4,
		BrightnessContrast: componentDefinition$5,
		Bulge: componentDefinition$6,
		ChannelBlur: componentDefinition$7,
		Checkerboard: componentDefinition$8,
		ChromaFlow: componentDefinition$9,
		ChromaticAberration: componentDefinition$10,
		Circle: componentDefinition$11,
		CRTScreen: componentDefinition$12,
		CursorTrail: componentDefinition$13,
		DiffuseBlur: componentDefinition$14,
		Dither: componentDefinition$15,
		DotGrid: componentDefinition$16,
		Duotone: componentDefinition$17,
		FilmGrain: componentDefinition$18,
		FloatingParticles: componentDefinition$19,
		GlassTiles: componentDefinition$20,
		Glow: componentDefinition$21,
		Godrays: componentDefinition$22,
		Grayscale: componentDefinition$23,
		Grid: componentDefinition$24,
		GridDistortion: componentDefinition$25,
		Group: componentDefinition$26,
		Halftone: componentDefinition$27,
		HueShift: componentDefinition$28,
		ImageTexture: componentDefinition$29,
		Invert: componentDefinition$30,
		LinearBlur: componentDefinition$31,
		LinearGradient: componentDefinition$32,
		Liquify: componentDefinition$33,
		Perspective: componentDefinition$34,
		Pixelate: componentDefinition$35,
		PolarCoordinates: componentDefinition$36,
		Posterize: componentDefinition$37,
		ProgressiveBlur: componentDefinition$38,
		RadialGradient: componentDefinition$39,
		RectangularCoordinates: componentDefinition$40,
		Ripples: componentDefinition$41,
		Saturation: componentDefinition$42,
		Sharpness: componentDefinition$43,
		Shatter: componentDefinition$44,
		SimplexNoise: componentDefinition$45,
		SineWave: componentDefinition$46,
		SolidColor: componentDefinition$47,
		Spherize: componentDefinition$48,
		Spiral: componentDefinition$49,
		Strands: componentDefinition$50,
		Stretch: componentDefinition$51,
		Stripes: componentDefinition$52,
		Swirl: componentDefinition$53,
		TiltShift: componentDefinition$54,
		Tint: componentDefinition$55,
		Tritone: componentDefinition$56,
		Twirl: componentDefinition$57,
		Vibrance: componentDefinition$58,
		WaveDistortion: componentDefinition$59,
		WebcamTexture: componentDefinition$60,
		ZoomBlur: componentDefinition$61
	};
	Object.entries(shaderDefinitions).forEach(([fileName, def]) => {
		const propsMetadata = {};
		if (def.props) Object.entries(def.props).forEach(([propName, propConfig]) => {
			if (propConfig.ui) propsMetadata[propName] = {
				ui: propConfig.ui,
				default: propConfig.default,
				description: propConfig.description
			};
		});
		registry[def.name] = {
			name: def.name,
			fileName,
			category: def.category || "Uncategorized",
			definition: def,
			propsMetadata
		};
	});
	return registry;
};
const shaderRegistry = createShaderRegistry();
function getAllShaders() {
	return Object.values(shaderRegistry);
}
function getShaderByName(name) {
	return shaderRegistry[name];
}
function getShadersByCategory(category) {
	return Object.values(shaderRegistry).filter((shader) => shader.category === category);
}
function getShaderCategories() {
	const categories = new Set(Object.values(shaderRegistry).map((shader) => shader.category));
	return Array.from(categories).sort();
}
export { getAllShaders, getShaderByName, getShaderCategories, getShadersByCategory, shaderRegistry };
