// shaders/dist/core/SimplexNoise-Ce79jgYM.js
import { r as transformColor } from "./transformations-YbhRK-rd.js";
import { t as createAnimatedTime } from "./time-BZqyVJXt.js";
import { t as mixColors } from "./colorMixing-CZPFmiT4.js";
import { clamp, exp, float, mx_noise_float, screenUV, vec2, vec3, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "SimplexNoise",
	category: "Base Layers",
	description: "Organic noise with animated movement",
	props: {
		colorA: {
			default: "#ffffff",
			description: "First color",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color A"
			}
		},
		colorB: {
			default: "transparent",
			description: "Second color",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color B"
			}
		},
		scale: {
			default: 2,
			description: "Pattern scale (higher = larger patterns)",
			ui: {
				type: "range",
				min: -2,
				max: 5,
				step: .1,
				label: "Scale"
			}
		},
		balance: {
			default: 0,
			description: "Balance between colors (negative = more colorB, positive = more colorA)",
			ui: {
				type: "range",
				min: -1,
				max: 1,
				step: .01,
				label: "Balance"
			}
		},
		contrast: {
			default: 0,
			description: "Pattern contrast (higher = sharper transitions)",
			ui: {
				type: "range",
				min: -2,
				max: 5,
				step: .1,
				label: "Contrast"
			}
		},
		seed: {
			default: 0,
			description: "Random seed for pattern variation",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Seed"
			}
		},
		speed: {
			default: 1,
			description: "Animation speed",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const uv$1 = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectCorrectedUV = vec2(uv$1.x.mul(aspect), uv$1.y);
		const animTime = createAnimatedTime(params, uniforms.speed);
		const pos = aspectCorrectedUV.mul(exp(uniforms.scale.uniform)).add(uniforms.seed.uniform);
		const noise = mx_noise_float(vec3(pos.x, pos.y, animTime.mul(.5)));
		const contrastMultiplier = uniforms.contrast.uniform.add(1);
		const k = clamp(noise.mul(contrastMultiplier).add(uniforms.balance.uniform).mul(.5).add(.5), 0, 1);
		return mixColors(uniforms.colorB.uniform, uniforms.colorA.uniform, k, float(0));
	}
};
var SimplexNoise_default = componentDefinition;
export { componentDefinition as n, SimplexNoise_default as t };
