// shaders/dist/core/Spiral-W4dALX1z.js
import { i as transformColorSpace, r as transformColor, s as transformPosition, t as colorSpaceOptions } from "./transformations-YbhRK-rd.js";
import { t as createAnimatedTime } from "./time-BZqyVJXt.js";
import { t as mixColors } from "./colorMixing-CZPFmiT4.js";
import { abs, atan, clamp, float, fract, fwidth, length, mix, screenUV, smoothstep, vec2, vec4, viewportSize } from "three/tsl";
const componentDefinition = {
	name: "Spiral",
	category: "Base Layers",
	description: "Rotating spiral pattern with animated movement",
	props: {
		colorA: {
			default: "#000000",
			transform: transformColor,
			description: "Background color",
			ui: {
				type: "color",
				label: "Color A"
			}
		},
		colorB: {
			default: "#ffffff",
			transform: transformColor,
			description: "Spiral stroke color",
			ui: {
				type: "color",
				label: "Color B"
			}
		},
		strokeWidth: {
			default: .5,
			description: "Thickness of spiral stroke",
			ui: {
				type: "range",
				min: 0,
				max: 2,
				step: .01,
				label: "Stroke Width"
			}
		},
		strokeFalloff: {
			default: 0,
			description: "Stroke losing width further from center",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Stroke Falloff"
			}
		},
		softness: {
			default: 0,
			description: "Color transition sharpness (0 = hard edge, 1 = smooth fade)",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Softness"
			}
		},
		speed: {
			default: 1,
			description: "Animation speed (negative values reverse direction)",
			ui: {
				type: "range",
				min: -3,
				max: 3,
				step: .1,
				label: "Speed"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "The center point of the spiral",
			ui: {
				type: "position",
				label: "Center"
			}
		},
		scale: {
			default: 1,
			description: "Scale factor for spiral bands (higher = more bands, lower = fewer bands)",
			ui: {
				type: "range",
				min: .1,
				max: 5,
				step: .1,
				label: "Scale"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const baseUV = screenUV;
		const aspect = viewportSize.x.div(viewportSize.y);
		const aspectCorrectedUV = vec2(baseUV.x.mul(aspect), baseUV.y);
		const centerPos = vec2(uniforms.center.uniform.x.mul(aspect), uniforms.center.uniform.y.oneMinus());
		const uv$1 = aspectCorrectedUV.sub(centerPos);
		const t = createAnimatedTime(params, uniforms.speed);
		const l = length(uv$1);
		const adjustedLength = l;
		const angle = atan(uv$1.y, uv$1.x).sub(t);
		const TWO_PI = float(6.283185307);
		const angleNormalised = angle.div(TWO_PI);
		const offset = adjustedLength.mul(uniforms.scale.uniform).add(angleNormalised);
		const shape = abs(fract(offset).sub(.5)).mul(2);
		const baseWidth = clamp(uniforms.strokeWidth.uniform, uniforms.strokeFalloff.uniform.mul(.005), 1);
		const falloffAmount = clamp(uniforms.strokeFalloff.uniform, 0, 1).mul(l).oneMinus();
		const width = baseWidth.mul(falloffAmount);
		const fw = fwidth(offset);
		const pixelSize = mix(float(4).sub(smoothstep(.05, .4, uniforms.strokeWidth.uniform.mul(2)).mul(smoothstep(.05, .4, uniforms.strokeWidth.uniform.oneMinus().mul(2))).mul(3)).mul(fw), fwidth(shape), clamp(fw, 0, 1));
		const res = smoothstep(width.sub(pixelSize).sub(uniforms.softness.uniform), width.add(pixelSize).add(uniforms.softness.uniform), shape);
		const mixedColor = mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, res, uniforms.colorSpace.uniform);
		return vec4(mixedColor.rgb, mixedColor.a);
	}
};
var Spiral_default = componentDefinition;
export { componentDefinition as n, Spiral_default as t };
