// shaders/dist/core/Swirl-CfsYsr4X.js
import { i as transformColorSpace, r as transformColor, t as colorSpaceOptions } from "./transformations-YbhRK-rd.js";
import { t as createAnimatedTime } from "./time-BZqyVJXt.js";
import { t as mixColors } from "./colorMixing-CZPFmiT4.js";
import { cos, screenUV, sin, smoothstep, vec2, vec4 } from "three/tsl";
const componentDefinition = {
	name: "Swirl",
	category: "Base Layers",
	description: "Flowing swirl pattern with multi-layered noise",
	props: {
		colorA: {
			default: "#1275d8",
			description: "Primary gradient color",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color A"
			}
		},
		colorB: {
			default: "#e19136",
			description: "Secondary gradient color",
			transform: transformColor,
			ui: {
				type: "color",
				label: "Color B"
			}
		},
		speed: {
			default: 1,
			description: "Flow animation speed",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed"
			}
		},
		detail: {
			default: 1,
			description: "Level of detail and intricacy in the swirl patterns",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Detail"
			}
		},
		blend: {
			default: 50,
			description: "Skew color balance toward A (lower values) or B (higher values)",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Blend"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms } = params;
		const uvCoords = screenUV;
		const t = createAnimatedTime(params, uniforms.speed);
		const detail = uniforms.detail.uniform;
		const freq1 = detail;
		const distort1 = vec2(uvCoords.x.add(sin(uvCoords.y.mul(freq1.mul(1.7)).add(t.mul(.8))).mul(.12).add(cos(uvCoords.x.mul(freq1.mul(.9)).sub(t.mul(.5))).mul(.05))), uvCoords.y.add(cos(uvCoords.x.mul(freq1.mul(1.3)).sub(t.mul(.6))).mul(.12).add(sin(uvCoords.y.mul(freq1.mul(1.1)).add(t.mul(.7))).mul(.05))));
		const pattern1 = sin(distort1.x.mul(freq1.mul(2.1)).add(distort1.y.mul(freq1.mul(1.8))).add(t.mul(.4)));
		const freq2 = detail.mul(2.1);
		const distort2 = vec2(distort1.x.add(cos(distort1.y.mul(freq2.mul(2.7)).sub(t.mul(.45))).mul(.07).add(sin(distort1.x.mul(freq2.mul(1.9)).add(t.mul(.6))).mul(.04))), distort1.y.add(sin(distort1.x.mul(freq2.mul(2.3)).add(t.mul(.65))).mul(.07).add(cos(distort1.y.mul(freq2.mul(1.6)).sub(t.mul(.4))).mul(.04))));
		const pattern2 = cos(distort2.x.mul(freq2.mul(1.4)).sub(distort2.y.mul(freq2.mul(1.9))).add(t.mul(.35)));
		const freq3 = detail.mul(3.7);
		const distort3 = vec2(distort2.x.add(sin(distort2.y.mul(freq3.mul(1.8)).add(t.mul(.85))).mul(.04).add(cos(distort2.x.mul(freq3.mul(1.3)).sub(t.mul(.55))).mul(.025)).add(sin(distort2.x.add(distort2.y).mul(freq3.mul(.7)).add(t.mul(.9))).mul(.02))), distort2.y.add(cos(distort2.x.mul(freq3.mul(1.6)).sub(t.mul(.75))).mul(.04).add(sin(distort2.y.mul(freq3.mul(1.1)).add(t.mul(.5))).mul(.025)).add(cos(distort2.x.add(distort2.y).mul(freq3.mul(.8)).sub(t.mul(.95))).mul(.02))));
		const pattern3 = sin(distort3.x.mul(freq3.mul(1.1)).add(distort3.y.mul(freq3.mul(1.5))).sub(t.mul(.55)));
		const combinedPattern = pattern1.mul(.45).add(pattern2.mul(.35)).add(pattern3.mul(.2));
		const blendBias = uniforms.blend.uniform.sub(50).mul(.006);
		const blendFactor = smoothstep(.3, .7, combinedPattern.mul(.5).add(.5).add(blendBias));
		const colorMix = mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, blendFactor, uniforms.colorSpace.uniform);
		const shimmer = sin(t.mul(2.5).add(combinedPattern.mul(8))).mul(.015).add(1);
		const finalColor = colorMix.mul(shimmer);
		return vec4(finalColor.rgb, finalColor.a);
	}
};
var Swirl_default = componentDefinition;
export { componentDefinition as n, Swirl_default as t };
