// shaders/dist/core/Tint-CpnitsHw.js
import { r as transformColor } from "./transformations-YbhRK-rd.js";
import { dot, max, mix, vec3, vec4 } from "three/tsl";
const componentDefinition = {
	name: "Tint",
	category: "Adjustments",
	description: "Apply a color tint to the image",
	requiresChild: true,
	props: {
		color: {
			default: "#ff8800",
			transform: transformColor,
			description: "Tint color",
			ui: {
				type: "color",
				label: "Tint Color"
			}
		},
		amount: {
			default: .5,
			description: "Tint amount (0 = no tint, 1 = full tint)",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Amount"
			}
		},
		preserveLuminosity: {
			default: true,
			description: "Preserve original brightness",
			ui: {
				type: "checkbox",
				label: "Preserve Luminosity"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode }) => {
		if (!childNode) {
			console.error("You must pass a child component into the Tint shader.");
			return vec4(0);
		}
		const tintColor = uniforms.color.uniform;
		const amount = uniforms.amount.uniform;
		const preserveLum = uniforms.preserveLuminosity.uniform;
		const rgb = vec3(childNode.r, childNode.g, childNode.b);
		const luminanceWeights = vec3(.299, .587, .114);
		const originalLum = dot(rgb, luminanceWeights);
		const tinted = mix(rgb, tintColor.rgb, amount);
		const tintedLum = dot(tinted, luminanceWeights);
		const lumPreserved = tinted.mul(originalLum.div(max(tintedLum, 1e-4)));
		return vec4(preserveLum.select(lumPreserved, tinted), childNode.a);
	}
};
var Tint_default = componentDefinition;
export { componentDefinition as n, Tint_default as t };
