// shaders/dist/core/Tritone-DzvSKVfR.js
import { i as transformColorSpace, r as transformColor, t as colorSpaceOptions } from "./transformations-YbhRK-rd.js";
import { t as mixColors } from "./colorMixing-CZPFmiT4.js";
import { convertToTexture, dot, screenUV, smoothstep, vec3, vec4 } from "three/tsl";
const componentDefinition = {
	name: "Tritone",
	category: "Adjustments",
	description: "Map colors to three tones: shadows, midtones, highlights",
	requiresRTT: true,
	requiresChild: true,
	props: {
		colorA: {
			default: "#ce1bea",
			transform: transformColor,
			description: "First color (used for shadows/darkest areas)",
			ui: {
				type: "color",
				label: "Color A (Shadows)"
			}
		},
		colorB: {
			default: "#2fff00",
			transform: transformColor,
			description: "Second color (used for midtones)",
			ui: {
				type: "color",
				label: "Color B (Midtones)"
			}
		},
		colorC: {
			default: "#ffff00",
			transform: transformColor,
			description: "Third color (used for highlights/brightest areas)",
			ui: {
				type: "color",
				label: "Color C (Highlights)"
			}
		},
		blendMid: {
			default: .5,
			description: "Midpoint position between the three colors",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Midpoint"
			}
		},
		colorSpace: {
			default: "linear",
			transform: transformColorSpace,
			description: "Color space for color interpolation",
			ui: {
				type: "select",
				options: colorSpaceOptions,
				label: "Color Space"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const inputColor = childTexture.sample(screenUV);
		const luminance = dot(inputColor.rgb, vec3(.299, .587, .114));
		const shadowToMid = smoothstep(uniforms.blendMid.uniform.sub(.25), uniforms.blendMid.uniform, luminance);
		const lowerBlend = mixColors(uniforms.colorA.uniform, uniforms.colorB.uniform, shadowToMid, uniforms.colorSpace.uniform);
		const midToHighlight = smoothstep(uniforms.blendMid.uniform, uniforms.blendMid.uniform.add(.25), luminance);
		return vec4(mixColors(lowerBlend, mixColors(uniforms.colorB.uniform, uniforms.colorC.uniform, midToHighlight, uniforms.colorSpace.uniform), smoothstep(uniforms.blendMid.uniform.sub(.1), uniforms.blendMid.uniform.add(.1), luminance), uniforms.colorSpace.uniform).rgb, inputColor.a);
	}
};
var Tritone_default = componentDefinition;
export { componentDefinition as n, Tritone_default as t };
