// shaders/dist/core/WaveDistortion-BA1PyqYH.js
import { t as applyEdgeHandling } from "./edges-Bd7GP4s2.js";
import { o as transformEdges } from "./transformations-YbhRK-rd.js";
import { t as createAnimatedTime } from "./time-BZqyVJXt.js";
import { PI, abs, convertToTexture, cos, float, fract, screenUV, sin, step, vec2, vec4 } from "three/tsl";
const componentDefinition = {
	name: "WaveDistortion",
	category: "Distortions",
	description: "Wave-based distortion with multiple waveform types",
	requiresRTT: true,
	requiresChild: true,
	props: {
		strength: {
			default: .3,
			description: "Distortion intensity",
			ui: {
				type: "range",
				min: 0,
				max: 1,
				step: .01,
				label: "Strength"
			}
		},
		frequency: {
			default: 1,
			description: "Number of bends/waves",
			ui: {
				type: "range",
				min: .1,
				max: 10,
				step: .1,
				label: "Frequency"
			}
		},
		speed: {
			default: 1,
			description: "Animation speed",
			ui: {
				type: "range",
				min: 0,
				max: 5,
				step: .1,
				label: "Speed"
			}
		},
		angle: {
			default: 0,
			description: "Direction of wave distortion in degrees",
			ui: {
				type: "range",
				min: 0,
				max: 360,
				step: 1,
				label: "Angle"
			}
		},
		waveType: {
			default: "sine",
			description: "Shape of the distortion wave",
			transform: (value) => {
				return {
					"sine": 0,
					"triangle": 1,
					"square": 2,
					"sawtooth": 3
				}[value] ?? 0;
			},
			ui: {
				type: "select",
				options: [
					{
						label: "Sine",
						value: "sine"
					},
					{
						label: "Triangle",
						value: "triangle"
					},
					{
						label: "Square",
						value: "square"
					},
					{
						label: "Sawtooth",
						value: "sawtooth"
					}
				],
				label: "Wave Type"
			}
		},
		edges: {
			default: "stretch",
			description: "How to handle edges when distortion pushes content out of bounds",
			transform: transformEdges,
			ui: {
				type: "select",
				options: [
					{
						label: "Stretch",
						value: "stretch"
					},
					{
						label: "Transparent",
						value: "transparent"
					},
					{
						label: "Mirror",
						value: "mirror"
					},
					{
						label: "Wrap",
						value: "wrap"
					}
				],
				label: "Edges"
			}
		}
	},
	fragmentNode: (params) => {
		const { uniforms, childNode, onCleanup } = params;
		if (!childNode) return vec4(0, 0, 0, 0);
		const uvCoord = screenUV;
		const t = createAnimatedTime(params, uniforms.speed).mul(.5);
		const angleRad = uniforms.angle.uniform.mul(PI).div(180);
		const cosAngle = cos(angleRad);
		const sinAngle = sin(angleRad);
		const centeredUV = uvCoord.sub(vec2(.5, .5));
		const phase = vec2(centeredUV.x.mul(cosAngle).sub(centeredUV.y.mul(sinAngle)), centeredUV.x.mul(sinAngle).add(centeredUV.y.mul(cosAngle))).add(vec2(.5, .5)).y.mul(uniforms.frequency.uniform).mul(PI.mul(2)).add(t);
		const normalizedPhase = fract(phase.div(PI.mul(2)));
		const sineWave = sin(phase);
		const triangleWave = abs(normalizedPhase.mul(2).sub(1)).mul(2).sub(1);
		const squareWave = step(float(.5), normalizedPhase).mul(2).sub(1);
		const sawtoothWave = normalizedPhase.mul(2).sub(1);
		const waveTypeValue = uniforms.waveType.uniform;
		const isSine = step(waveTypeValue, float(.5));
		const isTriangle = step(float(.5), waveTypeValue).mul(step(waveTypeValue, float(1.5)));
		const isSquare = step(float(1.5), waveTypeValue).mul(step(waveTypeValue, float(2.5)));
		const isSawtooth = step(float(2.5), waveTypeValue);
		const displacement = sineWave.mul(isSine).add(triangleWave.mul(isTriangle)).add(squareWave.mul(isSquare)).add(sawtoothWave.mul(isSawtooth)).mul(uniforms.strength.uniform).mul(.5);
		const displacementX = displacement.mul(cosAngle);
		const displacementY = displacement.mul(sinAngle);
		const distortedUV = vec2(uvCoord.x.add(displacementX), uvCoord.y.add(displacementY));
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		return applyEdgeHandling(distortedUV, childTexture.sample(distortedUV), childTexture, uniforms.edges.uniform);
	}
};
var WaveDistortion_default = componentDefinition;
export { componentDefinition as n, WaveDistortion_default as t };
