// shaders/dist/core/ZoomBlur-CIkkDBIo.js
import { s as transformPosition } from "./transformations-YbhRK-rd.js";
import { Fn, convertToTexture, float, screenUV, vec2, vec4 } from "three/tsl";
const componentDefinition = {
	name: "ZoomBlur",
	category: "Blurs",
	description: "Radial zoom blur expanding from a center point",
	requiresRTT: true,
	requiresChild: true,
	props: {
		intensity: {
			default: 30,
			description: "Intensity of the zoom blur effect",
			ui: {
				type: "range",
				min: 0,
				max: 100,
				step: 1,
				label: "Intensity"
			}
		},
		center: {
			default: {
				x: .5,
				y: .5
			},
			transform: transformPosition,
			description: "Center point of the zoom blur",
			ui: {
				type: "position",
				label: "Center"
			}
		}
	},
	fragmentNode: ({ uniforms, childNode, onCleanup }) => {
		if (!childNode) return vec4(0, 0, 0, 0);
		const childTexture = convertToTexture(childNode);
		onCleanup(() => {
			if (childTexture?.renderTarget?.dispose) childTexture.renderTarget.dispose();
		});
		const center = vec2(uniforms.center.uniform.x, uniforms.center.uniform.y.oneMinus());
		const radius = uniforms.intensity.uniform.mul(.01);
		return Fn(() => {
			const uv$1 = screenUV;
			const total = vec4(0).toVar();
			const totalWeight = float(0).toVar();
			for (let i = 0; i < 32; i++) {
				const t = float(i).div(float(31));
				const weightT = t.sub(.5).mul(2);
				const weight = float(1).div(float(2.506628)).mul(float(-.5).mul(weightT).mul(weightT).div(float(.64)).exp());
				const scale = float(1).add(radius.mul(t));
				const scaledCoord = uv$1.sub(center).div(scale).add(center);
				const sample = childTexture.sample(scaledCoord).mul(weight);
				total.assign(total.add(sample));
				totalWeight.assign(totalWeight.add(weight));
			}
			return total.div(totalWeight);
		})();
	}
};
var ZoomBlur_default = componentDefinition;
export { componentDefinition as n, ZoomBlur_default as t };
