// shaders/dist/solid/utils/generatePresetCode.js
const DEFAULT_TRANSFORM = {
  offsetX: 0,
  offsetY: 0,
  rotation: 0,
  scale: 1,
  anchorX: 0.5,
  anchorY: 0.5,
  edges: "transparent"
};
function toObjectLiteral(obj) {
  const entries = Object.entries(obj).map(([key, value]) => {
    const valueStr = typeof value === "string" ? `"${value}"` : value;
    return `${key}: ${valueStr}`;
  });
  return `{ ${entries.join(", ")} }`;
}
const shaderMetadata = {
  "AngularBlur": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Ascii": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Beam": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Blob": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Blur": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "BrightnessContrast": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Bulge": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "CRTScreen": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "ChannelBlur": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Checkerboard": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "ChromaFlow": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "ChromaticAberration": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Circle": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "CursorTrail": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "DiffuseBlur": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Dither": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "DotGrid": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Duotone": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "FilmGrain": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "FloatingParticles": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "GlassTiles": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Glow": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Godrays": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Grayscale": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Grid": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "GridDistortion": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Group": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Halftone": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "HueShift": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "ImageTexture": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Invert": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "LinearBlur": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "LinearGradient": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Liquify": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Perspective": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Pixelate": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "PolarCoordinates": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Posterize": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "ProgressiveBlur": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "RadialGradient": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "RectangularCoordinates": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Ripples": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Saturation": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Sharpness": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Shatter": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "SimplexNoise": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "SineWave": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "SolidColor": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Spherize": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Spiral": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Strands": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Stretch": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Stripes": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Swirl": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "TiltShift": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Tint": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Tritone": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Twirl": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "Vibrance": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "WaveDistortion": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "WebcamTexture": {
    "opacity": 1,
    "blendMode": "normal"
  },
  "ZoomBlur": {
    "opacity": 1,
    "blendMode": "normal"
  }
};
function generatePropString(props, componentType, indent = "  ") {
  return Object.entries(props).sort(([a], [b]) => a.localeCompare(b)).filter(([key, value]) => {
    if (key === "maskType" && value === "alpha") return false;
    if (key === "transform" && typeof value === "object") {
      const allDefaults = Object.keys(DEFAULT_TRANSFORM).every(
        (k) => value[k] === DEFAULT_TRANSFORM[k]
      );
      if (allDefaults) return false;
    }
    const componentDefaults = shaderMetadata[componentType] || {};
    if (componentDefaults.hasOwnProperty(key)) {
      const defaultValue = componentDefaults[key];
      if (value !== null && defaultValue !== null && typeof value === "object" && typeof defaultValue === "object") {
        return JSON.stringify(value) !== JSON.stringify(defaultValue);
      }
      return value !== defaultValue;
    }
    if (key === "opacity" && value === 1) return false;
    if (key === "blendMode" && value === "normal") return false;
    return true;
  }).map(([key, value]) => {
    if (key === "transform" && typeof value === "object") {
      const nonDefaultKeys = {};
      for (const k in value) {
        if (value[k] !== DEFAULT_TRANSFORM[k]) {
          nonDefaultKeys[k] = value[k];
        }
      }
      const keys = Object.keys(nonDefaultKeys).sort();
      if (keys.length > 1) {
        const entries = keys.map((k) => `${k}: ${typeof nonDefaultKeys[k] === "string" ? `"${nonDefaultKeys[k]}"` : nonDefaultKeys[k]}`);
        return `${key}={{
${indent}    ${entries.join(`,
${indent}    `)}
${indent}  }}`;
      } else if (keys.length === 1) {
        const k = keys[0];
        const v = nonDefaultKeys[k];
        return `${key}={{ ${k}: ${typeof v === "string" ? `"${v}"` : v} }}`;
      }
    }
    if (typeof value === "string") {
      return `${key}=${JSON.stringify(value)}`;
    } else if (value !== null && typeof value === "object") {
      const roundedValue = { ...value };
      if ("x" in roundedValue && typeof roundedValue.x === "number") {
        roundedValue.x = Math.round(roundedValue.x * 100) / 100;
      }
      if ("y" in roundedValue && typeof roundedValue.y === "number") {
        roundedValue.y = Math.round(roundedValue.y * 100) / 100;
      }
      return `${key}={${toObjectLiteral(roundedValue)}}`;
    } else {
      return `${key}={${value}}`;
    }
  }).join("\n" + indent + "  ");
}
function isIdUsedAsMaskSource(id, allComponents) {
  const flatComponents = [];
  function flattenComponents(components) {
    for (const component of components) {
      flatComponents.push(component);
      if (component.children) {
        flattenComponents(component.children);
      }
    }
  }
  flattenComponents(allComponents);
  return flatComponents.some(
    (component) => {
      var _a;
      return ((_a = component.props) == null ? void 0 : _a.maskSource) === id;
    }
  );
}
function generatePresetCode(preset) {
  const generateComponentString = (config, indent = "  ") => {
    const propString = config.props ? generatePropString(config.props, config.type, indent) : "";
    let idString = "";
    if (config.id) {
      const isUsedAsMask = isIdUsedAsMaskSource(config.id, preset.components);
      if (isUsedAsMask) {
        idString = `id="${config.id}"`;
      }
    }
    const attributes = [idString, propString].filter(Boolean).join("\n" + indent + "  ");
    if (config.children && config.children.length > 0) {
      const childrenString = config.children.map((child) => generateComponentString(child, indent + "  ")).join("\n");
      if (attributes) {
        return `${indent}<${config.type}
${indent}  ${attributes}>
${childrenString}
${indent}</${config.type}>`;
      } else {
        return `${indent}<${config.type}>
${childrenString}
${indent}</${config.type}>`;
      }
    } else {
      if (attributes) {
        return `${indent}<${config.type}
${indent}  ${attributes} />`;
      } else {
        return `${indent}<${config.type} />`;
      }
    }
  };
  const componentStrings = preset.components.map((config) => generateComponentString(config)).join("\n");
  return `<Shader>
${componentStrings}
</Shader>`;
}
const availableComponents = [
  // @ts-ignore - replaced at build time
  "AngularBlur",
  "Ascii",
  "Beam",
  "Blob",
  "Blur",
  "BrightnessContrast",
  "Bulge",
  "CRTScreen",
  "ChannelBlur",
  "Checkerboard",
  "ChromaFlow",
  "ChromaticAberration",
  "Circle",
  "CursorTrail",
  "DiffuseBlur",
  "Dither",
  "DotGrid",
  "Duotone",
  "FilmGrain",
  "FloatingParticles",
  "GlassTiles",
  "Glow",
  "Godrays",
  "Grayscale",
  "Grid",
  "GridDistortion",
  "Group",
  "Halftone",
  "HueShift",
  "ImageTexture",
  "Invert",
  "LinearBlur",
  "LinearGradient",
  "Liquify",
  "Perspective",
  "Pixelate",
  "PolarCoordinates",
  "Posterize",
  "ProgressiveBlur",
  "RadialGradient",
  "RectangularCoordinates",
  "Ripples",
  "Saturation",
  "Sharpness",
  "Shatter",
  "SimplexNoise",
  "SineWave",
  "SolidColor",
  "Spherize",
  "Spiral",
  "Strands",
  "Stretch",
  "Stripes",
  "Swirl",
  "TiltShift",
  "Tint",
  "Tritone",
  "Twirl",
  "Vibrance",
  "WaveDistortion",
  "WebcamTexture",
  "ZoomBlur"
];
export {
  availableComponents,
  generatePresetCode
};
